pkg/controller/flink/flink.go

```diff
@@ -348,52 +317,84 @@ func listToFlinkDeployment(ds []v1.Deployment, hash string) *common.FlinkDeploym
+	appHash := HashForApplication(app)
+
+	if appHash == app.Status.FailedDeployHash {
+		return app.Status.DeployHash
+	}
+	return appHash
+}
+
 // Gets the current deployment and any other deployments for the application. The current deployment will be the one
 // that matches the FlinkApplication, unless the FailedDeployHash is set, in which case it will be the one with that
 // hash.
-func (f *Controller) GetCurrentAndOldDeploymentsForApp(ctx context.Context,
-	application *v1alpha1.FlinkApplication) (*common.FlinkDeployment, []common.FlinkDeployment, error) {
-	appLabels := k8.GetAppLabel(application.Name)
-	deployments, err := f.k8Cluster.GetDeploymentsWithLabel(ctx, application.Namespace, appLabels)
+func (f *Controller) GetCurrentDeploymentsForApp(ctx context.Context, application *v1alpha1.FlinkApplication) (*common.FlinkDeployment, error) {
+	labels := k8.GetAppLabel(application.Name)
+	curHash := getCurrentHash(application)
+	labels[FlinkAppHash] = curHash
+
+	deployments, err := f.k8Cluster.GetDeploymentsWithLabel(ctx, application.Namespace, labels)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	byHash := map[string][]v1.Deployment{}
-	for _, deployment := range deployments.Items {
-		byHash[deployment.Labels[FlinkAppHash]] = append(byHash[deployment.Labels[FlinkAppHash]], deployment)
+	cur := listToFlinkDeployment(deployments.Items, curHash)
+	if cur != nil && application.Status.FailedDeployHash == "" &&
+		(!f.deploymentMatches(ctx, cur.Jobmanager, application) || !f.deploymentMatches(ctx, cur.Taskmanager, application)) {
+		// we had a hash collision (i.e., the previous application has the same hash as the new one)
+		// this is *very* unlikely to occur (1/2^32)
+		return nil, errors.New("found hash collision for deployment, you must do a clean deploy")
 	}
 
-	appHash := HashForApplication(application)
-	var curHash string
+	return cur, nil
+}
 
-	if appHash == application.Status.FailedDeployHash {
-		curHash = application.Status.DeployHash
-	} else {
-		curHash = appHash
+func (f *Controller) DeleteOldResourcesForApp(ctx context.Context, app *v1alpha1.FlinkApplication) error {
+	curHash := getCurrentHash(app)
+
+	appLabel := k8.GetAppLabel(app.Name)
+	deployments, err := f.k8Cluster.GetDeploymentsWithLabel(ctx, app.Namespace, appLabel)
+	if err != nil {
+		return err
 	}
 
-	cur := listToFlinkDeployment(byHash[curHash], curHash)
-	if cur != nil && application.Status.FailedDeployHash == "" &&
-		(!f.deploymentMatches(ctx, cur.Jobmanager, application) || !f.deploymentMatches(ctx, cur.Taskmanager, application)) {
-		// we had a hash collision (i.e., the previous application has the same hash as the new one)
-		// this is *very* unlikely to occur (1/2^32)
-		return nil, nil, errors.New("found hash collision for deployment, you must do a clean deploy")
+	oldObjects := make([]metav1.Object, 0)
+
+	for _, d := range deployments.Items {
+		if d.Labels[FlinkAppHash] != curHash {
```

Comment:
Can you also add extra check label map with has key `FlinkAppHash`, and also check if value is not empty string.